docarray/array/storage/base/getsetdel.py

```diff
@@ -173,15 +186,26 @@ def _set_docs_by_ids(self, ids, docs: Iterable['Document'], mismatch_ids: Dict):
         for _id, doc in zip(ids, docs):
             self._set_doc_by_id(_id, doc)
 
-    def _update_subindices_set(self, ids, docs):
-        if getattr(self, '_subindices', None):
-            for selector, da in self._subindices.items():
-                old_ids = DocumentArray(self[ids])[
-                    selector, 'id'
-                ]  # hack to get the Document['@c'] without having to do Document.chunks
-                with da:
-                    del da[old_ids]
-                    da.extend(DocumentArray(docs)[selector])  # same hack here
+    def _update_subindices_set(self, set_index, docs):
+        subindices = getattr(self, '_subindices', None)
+        if not subindices:
+            return
+        if isinstance(set_index, tuple):  # handled later in recursive call
+            return
+        if isinstance(set_index, str) and set_index.startswith('@'):
+            # 'nested' (non root-level) set, update entire subindex directly
+            _check_valid_values_nested_set(DocumentArray(self[set_index]), docs)
+            if set_index in subindices.keys():
+                subindex_da = subindices[set_index]
+                with subindex_da:
+                    subindices[set_index].clear()
+                    subindices[set_index].extend(docs)
+        else:  # root level set, update subindices iteratively
+            for subindex_selector, subindex_da in subindices.items():
+                old_ids = DocumentArray(self[set_index])[subindex_selector, 'id']
+                with subindex_da:
+                    del subindex_da[old_ids]
+                    subindex_da.extend(DocumentArray(docs)[subindex_selector])
```

Comment:
probably this is not the correct way to update the subindex but I'm afraid there is no better efficient way:
if you delete document chunks (the root document is positioned in a middle position) from the subindex and you append its new chunks to the subindex, we are changing the order of docs within the subindex, and I think order matters in some minor cases

Reply:
This is correct, we knew about this and decided that we don't want to make guarantees about the order in the subindex. If you care about it i think you can still get the order from the serialized chunks that are stored in the root index